### ISSUE TITLE
A better `no-nested-ternary` rule

### ISSUE THREAD

**Issue by sindresorhus** [sindresorhus/eslint-plugin-unicorn#50]
[<img alt="Issuehunt badges" src="https://issuehunt.io/r/sindresorhus/eslint-plugin-unicorn/issues/50/badge.svg" />](https://issuehunt.io/r/sindresorhus/eslint-plugin-unicorn/issues/50)

Since it was denied in ESLint: https://github.com/eslint/eslint/issues/3480

I think it would be better with a new rule that only errors on nested ternaries that are not wrapped in parens. For example: `parens-wrapped-nested-ternary` (a mouthful, but very descriptive).
## 

Thoughts?

> There is a $82.00 open bounty on this issue. Add more on [Issuehunt](https://issuehunt.io/r/sindresorhus/eslint-plugin-unicorn/issues/50).
> - Checkout the [Issuehunt explorer](https://issuehunt.io/r/sindresorhus/eslint-plugin-unicorn/) to discover more funded issues.
> - Need some help from other developers? [Add your repositories](https://issuehunt.io/r/new) on Issuehunt to raise funds.

**Comment by sindresorhus** [comment]
@Qix-

**Comment by SamVerschueren** [comment]
I agree that this should be allowed:

```js
var foo = i > 5 ? true : (i < 100 ? true : false);
```

But should we prevent nesting multiple levels?

```js
var foo = i > 5 ? true : (i < 100 ? true : (i < 1000 ? true : false));
```

Because that is IMO not readable at all. So I'm 👍 on nesting with parens, but only one level. Or maybe a configurable level that defaults to 1.

**Comment by sindresorhus** [comment]
Yes, forgot to mention that. Only one level. Shouldn't be configurable.

These should be allowed:

```js
var foo = i > 5 ? true : (i < 100 ? true : false);

var foo = i > 5 ? (i < 100 ? true : false) : true;

var foo = i > 5 ? (i < 100 ? true : false) : (i < 100 ? true : false);
```

-

Unless I'm missing something, this rule could be fixable, by just wrapping the nested ternary in parens.

**Comment by SamVerschueren** [comment]
👍 

> Unless I'm missing something, this rule could be fixable, by just wrapping the nested ternary in parens.

That would be great!

**Comment by jfmengels** [comment]
Sounds good to me. So we allow only up to 1 ternary expression directly inside a ternary expression.

We could forbid additional ternary expressions inside different expressions that are inside a ternary. An exemple is more descriptive:
```js
var foo = i > 5 ? true : (i < 100 ? true : false); // is fine

var foo = i > 5 ? true : (i < 100 ? (i > 50 ? false : true) : false); // Not fine, 3 levels of ternary expression

var foo = i > 5 ? true : (i < 100 ? FOO(i > 50 ? false : true) : false); // This one
```
Should we do the additional work of forbidding that last case? My feeling would be no, as it sounds pretty error-prone and not the core idea of the rule.

**Comment by sindresorhus** [comment]
> Should we do the additional work of forbidding that last case?

No

**Comment by sindresorhus** [comment]
Also open to a better rule name.

**Comment by jfmengels** [comment]
I'm thinking along the lines of no-confusing-nested-ternary.

**Comment by sindresorhus** [comment]
👍

**Comment by Qix-** [comment]
Ehh why not just `no-confusing-ternary`?

**Comment by jfmengels** [comment]
I'd be fine with `no-confusing-ternary`. Simply `unicorn/no-nested-ternary' would also work actually.

**Comment by kevva** [comment]
> Simply `unicorn/no-nested-ternary' would also work actually.

If wrapped in parens, wouldn't that be false? Thinking of the following example above:

```js
var foo = i > 5 ? (i < 100 ? true : false) : (i < 100 ? true : false);
```

**Comment by jfmengels** [comment]
Yeah you're right... I thought that we could just name it as the core rule as it's just allowing one "edge" case, but it might be more confusing that we then do accept that edge case.

**Comment by Qix-** [comment]
That would be confusing for sure. The rule name should definitely be different.

**Comment by sompylasar** [comment]
Please could anyone explain why the community is bothered by the outer parens around complex expressions around the right-hand-side of assignments?

How 
```
const foo = i > 5 ? <GreaterThanFive /> : (i < 100 ? <LessThanHundred /> : <GreaterThanHundred />);
``` 
is more readable than 
```
const foo = ( i > 5 ? <GreaterThanFive /> : (i < 100 ? <LessThanHundred /> : <GreaterThanHundred />) );
```
and  than 
```
const foo = ( i > 5
    ? <GreaterThanFive />
    : ( i < 100
        ? <LessThanHundred />
        : <GreaterThanHundred />
    )
);
```
?

The latter looks most similar to the `if`:
```
let foo;
if ( i > 5) {
  foo = <GreaterThanFive />;
}
else if ( i < 100 ) {
  foo = <LessThanHundred />;
}
else {
  foo = <GreaterThanHundred />;
}
```

**Comment by sindresorhus** [comment]
https://github.com/not-an-aardvark/eslint-rule-composer might let us just modify the built-in rule instead of making one from scratch.

**Comment by IssuehuntBot** [comment]
@0maxxam0 has funded $2.00 to this issue.

---
- Submit pull request via [IssueHunt](https://issuehunt.io/repos/55832243/issues/50) to receive this reward.
- Want to contribute? Chip in to this issue via [IssueHunt](https://issuehunt.io/repos/55832243/issues/50).
- Checkout the [IssueHunt Issue Explorer](https://issuehunt.io/issues) to see more funded issues.
- Need help from developers? [Add your repository](https://issuehunt.io/r/new) on IssueHunt to raise funds.

**Comment by zeorin** [comment]
Hmm, I'm late to the party here, but a different approach to using parens is to only allow nested ternaries in the alternate sub-expression, effectively creating a daisy chain.

My change to ESLint, too, was rejected: https://github.com/eslint/eslint/pull/10452

**Comment by IssuehuntBot** [comment]
@issuehunt has funded $80.00 to this issue.

---
- Submit pull request via [IssueHunt](https://issuehunt.io/repos/55832243/issues/50) to receive this reward.
- Want to contribute? Chip in to this issue via [IssueHunt](https://issuehunt.io/repos/55832243/issues/50).
- Checkout the [IssueHunt Issue Explorer](https://issuehunt.io/issues) to see more funded issues.
- Need help from developers? [Add your repository](https://issuehunt.io/r/new) on IssueHunt to raise funds.
